client/mm/exchange_adaptor.go

```diff
@@ -3303,6 +3366,146 @@ func (u *unifiedExchangeAdaptor) Book() (buys, sells []*core.MiniOrder, _ error)
 	return u.CEX.Book(u.baseID, u.quoteID)
 }
 
+func (u *unifiedExchangeAdaptor) problems() *BotProblems {
+	u.botProblemsMtx.RLock()
+	defer u.botProblemsMtx.RUnlock()
+	return u.botProblems.copy()
+}
+
+func (u *unifiedExchangeAdaptor) updateBotProblems(f func(*BotProblems)) {
+	u.botProblemsMtx.Lock()
+	defer u.botProblemsMtx.Unlock()
+	f(u.botProblems)
+}
+
+// tradingLimitNotReached returns true if the user has not reached their trading
+// limit.
+func (u *unifiedExchangeAdaptor) tradingLimitNotReached() bool {
+	var tradingLimitReached bool
+	var err error
+	defer u.updateBotProblems(func(problems *BotProblems) {
+		if err != nil {
+			problems.AdditionalError = err
+		}
+		problems.UserLimitTooLow = tradingLimitReached
+	})
+
+	userParcels, parcelLimit, err := u.clientCore.TradingLimits(u.host)
+	if err != nil {
+		return false
+	}
+
+	tradingLimitReached = userParcels >= parcelLimit
+	return !tradingLimitReached
+}
+
+// updateDepositWithdrawProblems updates BotProblems with the result of the
+// last deposit or withdrawal attempt for a specific asset.
+func (u *unifiedExchangeAdaptor) updateDepositWithdrawProblems(base, deposit bool, err error) {
+	u.updateBotProblems(func(problems *BotProblems) {
+		assetID := u.quoteID
+		if base {
+			assetID = u.baseID
+		}
+
+		if deposit {
+			if err == nil {
+				delete(problems.DepositErr, assetID)
+			} else {
+				problems.DepositErr[assetID] = newStampedError(err)
+			}
+		} else {
+			if err == nil {
+				delete(problems.WithdrawErr, assetID)
+			} else {
+				problems.WithdrawErr[assetID] = newStampedError(err)
+			}
+		}
+	})
+}
+
+// updateCEXTradeError updates BotProblems with the result of the last CEX trade
+// attempt.
+func (u *unifiedExchangeAdaptor) updateCEXTradeError(err error) {
+	u.updateBotProblems(func(problems *BotProblems) {
+		if err == nil {
+			problems.CEXTradeErr = nil
+		} else {
+			problems.CEXTradeErr = newStampedError(err)
+		}
+	})
+}
+
+// checkBotHealth returns true if the bot is healthy and can continue trading.
+func (u *unifiedExchangeAdaptor) checkBotHealth() bool {
+	user := u.clientCore.User()
```

Comment:
Pulling the entire `core.User` is way too much. Please find or create a more efficient way to get the little bit of data you need from core.